### PR TITLE
8303921: serviceability/sa/UniqueVtableTest.java timed out

### PR DESCRIPTION
**Pull request by alexmenkov**
The change:
- updates UniqueVtableTest to follow standard SA way - attach to target from subprocess and use SATestUtils.addPrivilegesIfNeeded for the subprocess;
- updates several tests in the same directory to resolve NoClassDefFoundError failures; It's known JTReg issue that "@build" actions for part of used shared classes may cause intermittent NoClassDefFoundError in other tests which use the same shared library classpath.

Tested: 100 runs on all platforms, no failures

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8303921](https://bugs.openjdk.org/browse/JDK-8303921): serviceability/sa/UniqueVtableTest.java timed out

### Reviewers
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/13030/head:pull/13030` \
`$ git checkout pull/13030`

Update a local copy of the PR: \
`$ git checkout pull/13030` \
`$ git pull https://git.openjdk.org/jdk pull/13030/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13030`

View PR using the GUI difftool: \
`$ git pr show -t 13030`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13030.diff">https://git.openjdk.org/jdk/pull/13030.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/jtreg/serviceability/sa/UniqueVtableTest.java [hunk 1]
```diff
@@ -170,9 +188,11 @@ private void run() throws Throwable {
...
         SATestUtils.skipIfCannotAttach(); // throws SkippedException if attach not expected to work.
 
-        UniqueVtableTest test = new UniqueVtableTest();
-
-        test.run();
+        if (args == null || args.length == 0) {
+            runMain();
+        } else {
+            runTest(Long.parseLong(args[0]));
+        }
```
Comment (plummercj) [inline test/hotspot/jtreg/serviceability/sa/UniqueVtableTest.java:195]: Could use some comments here. Also, I think `SATestUtils.skipIfCannotAttach` is only needed for the `else` part.

File: test/hotspot/jtreg/serviceability/sa/UniqueVtableTest.java [hunk 2]
```diff
@@ -131,26 +144,31 @@ private void runTest() throws Throwable {
...
+    }
+
+    private static void runMain() throws Throwable {
         Throwable reasonToFail = null;
         LingeredApp app = null;
         try {
             app = LingeredApp.startApp();
-            attach(app.getPid());
-            runTest();
+            createAnotherToAttach(app.getPid());
```
Comment (plummercj) [inline test/hotspot/jtreg/serviceability/sa/UniqueVtableTest.java:168]: Did you ever figure out why attaching from the main test process sometimes fails?

File: test/hotspot/jtreg/serviceability/sa/UniqueVtableTest.java [hunk 3]
```diff
@@ -131,26 +144,31 @@ private void runTest() throws Throwable {
...
+        ProcessBuilder processBuilder = ProcessTools.createJavaProcessBuilder(
+            "--add-modules=jdk.hotspot.agent",
+            "--add-exports=jdk.hotspot.agent/sun.jvm.hotspot=ALL-UNNAMED",
+            "--add-exports=jdk.hotspot.agent/sun.jvm.hotspot.debugger=ALL-UNNAMED",
+            "--add-exports=jdk.hotspot.agent/sun.jvm.hotspot.types=ALL-UNNAMED",
+            "--add-exports=jdk.hotspot.agent/sun.jvm.hotspot.types.basic=ALL-UNNAMED",
+            "UniqueVtableTest",
+            Long.toString(lingeredAppPid));
+        SATestUtils.addPrivilegesIfNeeded(processBuilder);
+        OutputAnalyzer SAOutput = ProcessTools.executeProcess(processBuilder);
```
Comment (plummercj) [inline test/hotspot/jtreg/serviceability/sa/UniqueVtableTest.java:158]: `SAOutput`: local variables should start with lower case.

**Review by plummercj — CHANGES_REQUESTED** [review]

File: test/hotspot/jtreg/serviceability/sa/UniqueVtableTest.java [hunk 4]
```diff
@@ -31,7 +31,7 @@
  *          jdk.hotspot.agent/sun.jvm.hotspot.types
  *          jdk.hotspot.agent/sun.jvm.hotspot.types.basic
  *
- * @run main/othervm UniqueVtableTest
+ * @run driver UniqueVtableTest
```
Comment (dcubed-ojdk) [inline test/hotspot/jtreg/serviceability/sa/UniqueVtableTest.java:34]: The other tests you touched in this PR use:
`@run main/othervm ...`
so why did this one have to change to:
`@run driver ...`

Comment (alexmenkov) [reply test/hotspot/jtreg/serviceability/sa/UniqueVtableTest.java:34]: Due changes in the test it doesn't need to be run in "othervm" mode, "driver" is ok now to make the test a bit faster
I didn't change mode other for other tests

File: test/hotspot/jtreg/serviceability/sa/UniqueVtableTest.java (hunk 2 above)
Comment (alexmenkov) [reply test/hotspot/jtreg/serviceability/sa/UniqueVtableTest.java:168]: No. It looks like JTReg executes main test process differently than the sub-process is executed, but I didn't find difference which can cause attach timeouts.

File: test/hotspot/jtreg/serviceability/sa/UniqueVtableTest.java (hunk 1 above)
Comment (alexmenkov) [reply test/hotspot/jtreg/serviceability/sa/UniqueVtableTest.java:195]: "else" part is a sub-process.
As far as I understand it SATestUtils.skipIfCannotAttach can be skipped for "else", but it's needed for main process.

Comment (alexmenkov) [reply test/hotspot/jtreg/serviceability/sa/UniqueVtableTest.java]: Added comment.
Left SATestUtils.skipIfCannotAttach as is (this is consistent with other SA tests)

File: test/hotspot/jtreg/serviceability/sa/UniqueVtableTest.java (hunk 3 above)
Comment (alexmenkov) [reply test/hotspot/jtreg/serviceability/sa/UniqueVtableTest.java]: Fixed

File: test/hotspot/jtreg/serviceability/sa/UniqueVtableTest.java (hunk 1 above)
Comment (plummercj) [reply test/hotspot/jtreg/serviceability/sa/UniqueVtableTest.java]: > "else" part is a sub-process.

Ok, I read it backwards. Part of the reason I asked for comments. :)

**Comment by dholmes-ora** [comment]
Not sure removing the build directives was the right way to go. As per the jtreg tag guide:

> A test that relies upon library classes should contain appropriate @build directives to ensure that the classes will be compiled. It is strongly recommended that tests do not rely on the use of implicit compilation by the Java compiler. 

so the problem is likely caused by missing build directives in the test(s) that fails.

**Review by sspitsyn — COMMENTED** [review]
This looks pretty good.
How did you test the fix?
Does it never fail with your fix?
Thanks,
Serguei

**Comment by alexmenkov** [comment]
> Not sure removing the build directives was the right way to go. As per the jtreg tag guide:
> 
> > A test that relies upon library classes should contain appropriate @build directives to ensure that the classes will be compiled. It is strongly recommended that tests do not rely on the use of implicit compilation by the Java compiler.
> 
> so the problem is likely caused by missing build directives in the test(s) that fails.

This is long standing issue with NoClassDefFoundError for library classes.
As far as I got from reading similar issues there was a try to add build directive for failing tests, but other tests from the same directory started to fail with NoClassDefFoundError later, and now most of the test have no build action for libraries.
It seems to me that it's much simpler to remove build action from 4 tests in the directory than add it for other 55

**Comment by alexmenkov** [comment]
> This looks pretty good. How did you test the fix? Does it never fail with your fix? Thanks, Serguei

I run test jobs for "serviceability/sa" 100 times on all platforms, not failures (neither attach timeout nor NoClassDefFoundError)

**Comment by dholmes-ora** [comment]
> It seems to me that it's much simpler to remove build action from 4 tests in the directory than add it for other 55

True. Sadly we keep getting bitten over and over by CODETOOLS-7902847. Sometimes the "fix" is to remove build directives (in Hotspot we switched from adding to removing back in 2017) and sometimes it is to add them (there have been some recent fixes that have done this - including by me).

**Review by sspitsyn — APPROVED** [review]

**Review by plummercj — APPROVED** [review]

**Comment by dcubed-ojdk** [comment]
Sigh... And again we have the situation where some folks are adding
`@build` directives and other folks are removing `@build` directives.

Another recent PR added library build directives:
https://github.com/openjdk/jdk/pull/13085

based on quoted guidance from the JTREG documentation.

This mess is related to:
[CODETOOLS-7902847](https://bugs.openjdk.org/browse/CODETOOLS-7902847) 
Class directory of a test case should not be used to compile a library

and NoClassDefFoundErrors show up when doing parallel execution of tests
where more than one test uses the "offending" library.

We really, really need @jonathan-gibbons to chime in on review threads like these.

**Pull request closed without merging**